### ISSUE TITLE
Runtime`PatternMatchFail` using `coalesce` when joining to subquery

### ISSUE THREAD

**Issue by amitojps** [bitemyapp/esqueleto#230]
This is on `esqueleto-3.3.3.0` and `persistent-2.10.4`. The exact error is

```
       uncaught exception: PatternMatchFail
       src/Database/Esqueleto/Internal/Internal.hs:2218:28-68: Non-exhaustive patterns in lambda
```

Here's the smallest example I've been able to construct. 

Models:
```
Foo
  Id UUID

Bar
  Id UUID
  fooId FooId
  field1 Text
  field2 Text Maybe
```

Code:
```
subquery :: SqlSetOperation (SqlExpr (Value FooId), SqlExpr (Value Text), SqlExpr (Value (Maybe Text)))
subquery = SelectQuery $ do
  bar <- from $ Table @Bar
  pure (bar ^. BarFooId, bar ^. BarField1, bar ^. BarField2)

workingQuery :: MonadUnliftIO m => ReaderT SqlBackend m [(Value FooId, Value Text, Value (Maybe Text))]
workingQuery =
  select $
    from $
      SelectQuery $ do
        (_ :& (fooId, txt, mTxt)) <-
          from $
            Table @Foo
            `InnerJoin` subquery
            `on` (\(foo :& (fooId, _ , _)) -> foo ^. FooId ==. fooId)
        pure (fooId, txt, mTxt)

failingQuery :: MonadUnliftIO m => ReaderT SqlBackend m [Value (Maybe Text)]
failingQuery =
  select $
    from $
      SelectQuery $ do
        (_ :& (_, txt, mTxt)) <-
          from $
            Table @Foo
            `InnerJoin` subquery
            `on` (\(foo :& (fooId, _ , _)) -> foo ^. FooId ==. fooId)
        pure (coalesce [mTxt, just txt])
```

Replacing this last line with things like `coalesceDefault [mTxt] txt` produces the same error.

**Comment by friedbrice** [comment]
I'm just gonna slide on in here so i get notifications on this topic.

**Comment by belevy** [comment]
Does this issue occur on v3.4.0.0?

**Comment by amitojps** [comment]
Good question, I didn't even realize 3.3.3.0 wasn't the most recent version! The issue is resolved on 3.4.0.0 and 3.4.0.1, thanks 👍

**Comment by parsonsmatt** [comment]
I'm going to go ahead and deprecate 3.3.0.0 so this will hopefully stop being less of a thing :sweat_smile:
